pkg/identity/cache/allocator_utils.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/sirupsen/logrus"
+)
+
+type UpdateIDsFunc func(added, deleted IdentityCache)
+type ReleaseFunc func(ctx context.Context, id *identity.Identity, notifyOwner bool) (released bool, err error)
+type GetLocalIDsFunc func() map[identity.NumericIdentity]*identity.Identity
+
+var UnknownIdentity = identity.NewIdentity(identity.IdentityUnknown, labels.Labels{labels.IDNameUnknown: labels.NewLabel(labels.IDNameUnknown, "", labels.LabelSourceReserved)})
+
+func WithholdLocalIdentities(localIDs, localNodeIDs *LocalIdentityCache, nids []identity.NumericIdentity) {
+	log.WithField(logfields.Identity, nids).Debug("Withholding numeric identities for later restoration")
+
+	// The allocators will return any identities that are not in-scope.
+	nids = localIDs.withhold(nids)
+	nids = localNodeIDs.withhold(nids)
+	if len(nids) > 0 {
+		log.WithField(logfields.Identity, nids).Error("Attempt to restore invalid numeric identities.")
+	}
+}
+
+func UnwithholdLocalIdentities(localIDs, localNodeIDs *LocalIdentityCache, nids []identity.NumericIdentity) {
+	log.WithField(logfields.Identity, nids).Debug("Unwithholding numeric identities")
+
+	// The allocators will ignore any identities that are not in-scope.
+	localIDs.withhold(nids)
+	localNodeIDs.withhold(nids)
+}
+
+func CompleteIdentityAllocation(err error, id *identity.Identity, allocated, isNewLocally, notifyOwner bool, updateIDsFunc UpdateIDsFunc) {
+	if err != nil {
+		return
+	}
+
+	if allocated || isNewLocally {
+		if id.ID.HasLocalScope() {
+			metrics.Identity.WithLabelValues(identity.NodeLocalIdentityType).Inc()
+		} else if id.ID.IsReservedIdentity() {
+			metrics.Identity.WithLabelValues(identity.ReservedIdentityType).Inc()
+		} else {
+			metrics.Identity.WithLabelValues(identity.ClusterLocalIdentityType).Inc()
+		}
+	}
+
+	if allocated && notifyOwner {
+		added := IdentityCache{
+			id.ID: id.LabelArray,
+		}
+		updateIDsFunc(added, nil)
+	}
+}
+
+func ResolveReservedIdentity(lbls labels.Labels, oldNID identity.NumericIdentity) (id *identity.Identity, allocated bool, err error) {
+	if option.Config.Debug {
+		log.WithFields(logrus.Fields{
+			logfields.IdentityLabels: lbls.String(),
+		}).Debug("Resolving identity")
+	}
+
+	// If there is only one label with the "reserved" source and a well-known
+	// key, use the well-known identity for that key.
+	if reservedIdentity := identity.LookupReservedIdentityByLabels(lbls); reservedIdentity != nil {
+		if option.Config.Debug {
+			log.WithFields(logrus.Fields{
+				logfields.Identity:       reservedIdentity.ID,
+				logfields.IdentityLabels: lbls.String(),
+				"isNew":                  false,
+			}).Debug("Resolved reserved identity")
+		}
+		return reservedIdentity, false, nil
+	}
+
+	return nil, false, nil
+}
+
+func CompleteRelease(owner IdentityAllocatorOwner, id *identity.Identity, released, notifyOwner bool, updateIDsFunc func(added, deleted IdentityCache)) {
+	defer func() {
```

Comment:
Only having a `defer()` function inside of a function doesn't make sense, the function will just immediately run the code inside the defer. Was the defer accidentally dropped from somewhere else?

Reply:
Yes, it was accidentally included here while refactoring. Thanks for noticing it!